File: assignment 9/python/programs/4.py
def find_max_length_subarray_with_sum(arr, target_sum):
    max_length = 0
    prefix_sum = 0
    sum_indices = {0: -1}

    for i, num in enumerate(arr):
        prefix_sum += num

        if prefix_sum - target_sum in sum_indices:
            subarray_length = i - sum_indices[prefix_sum - target_sum]
            if subarray_length > max_length:
                max_length = subarray_length

        if prefix_sum not in sum_indices:
            sum_indices[prefix_sum] = i

    return max_length


# Example usage
arr = [1, -1, 5, -2, 3]
target_sum = 3
max_length = find_max_length_subarray_with_sum(arr, target_sum)
print("Maximum length of subarray with sum", target_sum, "is", max_length)

File: assignment 9/c/programs/4.c
#include <stdio.h>

void findMaxSubarrayWithSum(int arr[], int size, int targetSum)
{
   int maxLength = 0;
   int currentSum = 0;
   int end = -1;
   int startIndex = 0;

   for (int i = 0; i < size; i++)
   {
      currentSum += arr[i];

      if (currentSum == targetSum)
      {
         maxLength = i + 1;
         end = i;
      }

      if (currentSum - targetSum >= 0)
      {
         int startIndexTemp = startIndex;
         while (currentSum - targetSum >= 0)
         {
            currentSum -= arr[startIndexTemp];
            startIndexTemp++;
         }

         if (i - startIndexTemp + 1 > maxLength)
         {
            maxLength = i - startIndexTemp + 1;
            end = i;
         }
         startIndex = startIndexTemp;
      }
   }

   if (maxLength > 0)
   {
      printf("Maximum length subarray with sum %d is [%d, %d]\n", targetSum, startIndex, end);
   }
   else
   {
      printf("No subarray with sum %d found.\n", targetSum);
   }
}

int main()
{
   int arr[] = {4, 2, -3, 1, 6};
   int size = sizeof(arr) / sizeof(arr[0]);
   int targetSum = 6;

   findMaxSubarrayWithSum(arr, size, targetSum);

   return 0;
}
